### PR TITLE
RFC: TPU SavedModel Export API for TF2

### PR DESCRIPTION
**Pull request by lzr-google**
**Review period open through 2019-11-23**

# TPU SavedModel Export API for TF2.x

| Status        | Proposed                                                |
| :------------ | :------------------------------------------------------ |
| **Author(s)** | Zhuoran Liu (lzr@google.com), Youlong Cheng (ylc@google.com) |
| **Sponsor**   | Jonathan Hseu (jhseu@google.com)                        |
| **Updated**   | 2019-11-06                                              |

## Objective

Provide an API to allow TF2 users to export TPU saved models <b>for
inference</b>, which:

+   Provide a user-friendly way to specify which function to run on TPU;
+   Hides Graph construction and TPU inference specific logic (multi-core
    support, etc) from users;
+   Allows specifying tags in SavedModel.

### Limitation of current `tf.saved_model.save()`

MetaGraphDef allows saving customized tags. Current downstream components like
TPU model-server, TFX infra-validator use the tags to load the specific
MetaGraph. However tf.saved_model.save() does not allow users to specify the set
of tags in MetaGraphDef, but hard-coded the MetaGraph to have only one ‘serve’
tag.

### CODE REVIEW

File: rfcs/20191106-tf2-tpu-savedmodel.md [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+signatures. The only way to achieve this is to hide those logics under
+`tf.keras.models.save_model()`.
+
+After the change, `tf.keras.models.save_model()` will have two additional
+arguments:
+
+1.  `export_to_tpu`: Simply setting this to `True` will export TPU model;
+2.  `tags_signatures`: Optionally for advanced users, if they want to have more
+    control of what tags / signatures they are using, they can use this argument
+    as if they are using TF2.x saving API.
```
[the remark this answers is not in the document]
Comment (lzr-google) [reply rfcs/20191106-tf2-tpu-savedmodel.md]: Sorry 2 was a mistake. It should be `tags`.
For 1, what suggestions do you have on how to let this method know we are exporting to TPU model?

File: rfcs/20191106-tf2-tpu-savedmodel.md [hunk 2]
```diff
@@ -0,0 +1,333 @@
+# TPU SavedModel Export API for TF2.x
+
+| Status        | Proposed                                                |
+| :------------ | :------------------------------------------------------ |
+| **RFC #**     | [NNN](https://github.com/tensorflow/community/pull/NNN) |
```
Comment (ewilderj) [inline rfcs/20191106-tf2-tpu-savedmodel.md]:
```suggestion
| **RFC #**     | [171](https://github.com/tensorflow/community/pull/171) |
```

Comment (lzr-google) [reply rfcs/20191106-tf2-tpu-savedmodel.md]: Nice catch. Thanks!

File: rfcs/20191106-tf2-tpu-savedmodel.md [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+    Node is done by attaching `_tpu_replicate` attribute;
++   For multicore efficiency, TPU computations have to be encapsulated as a
+    function and saved in FunctionLibrary, and will be called by
+    TPUPartitionedCall. A TPUOrdinalSelector node has to be connected to
+    TPUPartitionedCall to do efficient round-robin core selection;
++   Variable nodes have to be lifted from TPU functions, rewritten as
+    VarHandleOp, and consumed by ReadVariableOp.
+
+Also for reducing the number of TPU compilation, serving platforms(For example,
+Servomatic) prefers batching the inference requests with a few allowed batch
```
Comment (ewilderj) [inline rfcs/20191106-tf2-tpu-savedmodel.md]:
```suggestion
[TensorFlow Serving](https://www.tensorflow.org/tfx/guide/serving)) prefers batching the inference requests with a few allowed batch
```

File: rfcs/20191106-tf2-tpu-savedmodel.md [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+
+Keras users only need to pass `export_to_tpu=True` to save to TPU SavedModel.
+(Currently, we require the Keras model being saved to be completely
+TPU-compatible.)
+
+```python
+tf.keras.models.save_model(
+    model,
+    filepath='...',
+    export_to_tpu=True)
```
[the remark this answers is not in the document]
Comment (k-w-w) [reply rfcs/20191106-tf2-tpu-savedmodel.md]: If it's possible on the dist strat side, I prefer this idea over adding new arguments to the function.

[the remark this answers is not in the document]
Comment (jpienaar) [reply rfcs/20191106-tf2-tpu-savedmodel.md]: +1 hardcoding one particular architecture into save_model seems unfortunate and seems to be encoding device specific information at  a very high level (also then why not a export_to_gpu, export_to_npu, export_to_ipu, ...)

**Comment by brijk7** [comment]
@ewilderj : what are the next steps on this RFC? Looks like the comment period is over, and the design review is done as well.

**Comment by ewilderj** [comment]
Public design review meeting notes should be posted to this PR, and the PR updated to change status to "Approved" (assuming it was approved.) Once that's done, can be merged.

**Comment by brijk7** [comment]
@lzr-google : can you please provide an update from the design review?

**Comment by lzr-google** [comment]
Hi, we didn't reach full agreement on the proposed design in its current state. More work will be needed before we have a conclusion.
Thanks!

**Comment by lzr-google** [comment]
Update on the status of this RFC: After changing the design according to review feedback, the new design as in the latest commit has been accepted, with change checked in at https://github.com/tensorflow/tensorflow/commit/ee1dcbbd66c88696d47e27383d6b87b45092e6da .

**Pull request merged**